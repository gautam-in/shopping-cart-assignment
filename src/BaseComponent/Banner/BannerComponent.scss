/******************* BannerComponent *****************************/
@import '../../Styles/variable.scss';
@import '../../Styles/color-variable.scss';

.banner {
    position: relative;
    margin-bottom: $SMALL_MARGIN;
    box-shadow: 0px 10px 5px -5px $LIGHT_GREY;

    @mixin basic-btn {
        position: absolute;
        padding: $SMALL_PADDING;
        top: 50%;
        background-color: $TRANSPARENT_BLACK;
        border: none;
        color: white;
        cursor: pointer;
    }

    & .banner_carousel {
        & .banner_carousel_image {
            width: 100%;
            height: auto;
        }
    }

    & .banner_prev {
        @include basic-btn;
        left: 0;
    }

    & .banner_next {
        @include basic-btn;
        right: 0;
    }

    & .banner_dot-slider {
        display: flex;
        justify-content: center;

        & .banner_dot {
            width: 10px;
            height: 10px;
            border-radius: 50%;
            background-color: $LIGHT_GREY;
            display: inline;
            margin: 10px;
            border: none;
            cursor: pointer;
        }

        & .banner_dot--selected {
            background-color: $APP_BLACK;
        }
    }
}
