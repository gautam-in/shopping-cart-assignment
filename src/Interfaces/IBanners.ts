export default interface IBanners {
    id: string;
    order: number;
    isActive: boolean;
    bannerImageUrl: string;
    bannerImageAlt: string;
}
