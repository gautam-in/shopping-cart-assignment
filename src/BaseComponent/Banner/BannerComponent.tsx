import "./BannerComponent.scss";
import IBanners from "../../Interfaces/IBanners";
import { useState } from "react";
import classNames from "classnames";

interface IBannerComponentProps {
    bannerList: Array<IBanners>;
}

export const BannerComponent: React.FC<IBannerComponentProps> = (
    props: IBannerComponentProps
) => {
    const { bannerList } = props;
    const BASE_CLASSNAME: string = "banner";
    const CAROUSEL_CLASSNAME: string = `${BASE_CLASSNAME}_carousel`;
    const CAROUSEL_IMAGE_CLASSNAME: string = `${CAROUSEL_CLASSNAME}_image`;
    const PREV_BUTTON_CLASSNAME: string = `${BASE_CLASSNAME}_prev`;
    const NEXT_BUTTON_CLASSNAME: string = `${BASE_CLASSNAME}_next`;
    const DOT_BUTTON_CONTAINER_CLASSNAME: string = `${BASE_CLASSNAME}_dot-slider`;

    const [selectedBanner, setSelectedBanner] = useState<number>(0);

    enum btnType {
        PREV,
        NEXT,
    }

    const handleBannerChange = (type: btnType) => {
        let index = selectedBanner;
        if (type === btnType.PREV) {
            if (index > 0) {
                setSelectedBanner(--index);
            }
        }
        if (type === btnType.NEXT) {
            if (index < bannerList.length-1) {
                setSelectedBanner(++index);
            }
        }
    };

    const getSelectedClassName = (index: number) => {
        return classNames({
            "banner_dot": true,
            "banner_dot--selected": index === selectedBanner,
        });
    };

    return bannerList.length > 0 ? (
        <div className={BASE_CLASSNAME}>
            <div className={CAROUSEL_CLASSNAME}>
                <img
                    className={CAROUSEL_IMAGE_CLASSNAME}
                    src={bannerList[selectedBanner].bannerImageUrl}
                    alt={bannerList[selectedBanner].bannerImageAlt}
                />
            </div>
            <button
                className={PREV_BUTTON_CLASSNAME}
                onClick={() => handleBannerChange(btnType.PREV)}
            >
                PREV
            </button>
            <button
                className={NEXT_BUTTON_CLASSNAME}
                onClick={() => handleBannerChange(btnType.NEXT)}
            >
                NEXT
            </button>
            <div className={DOT_BUTTON_CONTAINER_CLASSNAME}>
                {bannerList.map((banner, index) => {
                    return (
                        <button
                            className={getSelectedClassName(index)}
                            onClick={() => setSelectedBanner(index)}
                        />
                    );
                })}
            </div>
        </div>
    ) : (
        <div>Something went wrong</div>
    );
};
