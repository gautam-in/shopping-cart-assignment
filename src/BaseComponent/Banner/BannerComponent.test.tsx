import { ShallowWrapper, shallow } from "enzyme";
import IBanners from "../../Interfaces/IBanners";
import { BannerComponent, IBannerComponentProps } from "./BannerComponent";

describe("BannerComponent", () => {
    const banner: IBanners = {
        id: "0",
        order: 0,
        isActive: true,
        bannerImageUrl: "",
        bannerImageAlt: "",
    };
    const bannerListProps: Array<IBanners> = [banner];
    const tree: ShallowWrapper<IBannerComponentProps> = shallow(
        <BannerComponent bannerList={bannerListProps} />
    );

    test("BannerComponent Snapshot render", () => {
        expect(tree).toMatchSnapshot();
    });

    test("Do Not Render Component if Banner List is Null", () => {
        const tree: ShallowWrapper<IBannerComponentProps> = shallow(
            <BannerComponent bannerList={[]} />
        );
        expect(tree.isEmptyRender()).toEqual(true);
    });

    test("Render Component if Banner List is Not Null", () => {
        const tree: ShallowWrapper<IBannerComponentProps> = shallow(
            <BannerComponent bannerList={bannerListProps} />
        );
        expect(tree.hasClass('banner')).toEqual(true);
    });
});
