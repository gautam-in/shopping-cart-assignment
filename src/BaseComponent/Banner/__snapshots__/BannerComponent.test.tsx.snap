// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BannerComponent BannerComponent Snapshot render 1`] = `ShallowWrapper {}`;
